File: 2020/9.jl
include("utils.jl")
using .AoCUtils: readInput
using Combinatorics

function loadInput()
    filename = "input_9.txt"
    lines = readInput(filename)
    numbers = parse.(Int, lines)
    return numbers
end

function XMAS(coding::Vector{Int}, n::Int)
    for c in combinations(coding, 2)
        if sum(c) == n
            return true
        end
    end
    return false
end

function part1(numbers::Vector{Int}, preamble::Int)
    for idx in preamble+1:length(numbers)
        if !XMAS(numbers[idx-preamble:idx-1], numbers[idx])
            println("🎄 Part 1: $(numbers[idx])")
            return numbers[idx]
        end
    end
end

function part2(numbers::Vector{Int}, target::Int)
    start = 1
    stop = 2
    sums = sum(numbers[start:stop])
    while sums != target
        if sums < target
            stop += 1
        elseif sums > target
            start += 1
        end
        sums = sum(numbers[start:stop])
    end
    res = minimum(numbers[start:stop]) + maximum(numbers[start:stop])
    println("🎄🎅 Part 2: $(res)")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 9         ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()
@time target = part1(inputs, 25)
@time part2(inputs, target)

File: 2020/9.py
import time
from utils import readInput
from itertools import combinations

def loadInput():
    lines = readInput("input_9.txt")    
    return [int(n) for n in lines]

def part1(numbers):
    
    step = 26
    for i in range(len(numbers)-step):
        code = numbers[i:i+step]
        sums = ([code[c[0]] + code[c[1]] for c in combinations(range(step-1), 2)])
        if code[-1] not in sums:
            print ("🎄 Part 1: {}".format(code[-1]))
            break
            

def part2(numbers):
    target = 10884537
    start = 0
    end = 1
    sums = sum(numbers[start:end])
    while sums != target:
        if sums < target:
            end += 1
        elif sums > target:
            start += 1
        sums = sum(numbers[start:end])
    print ("🎄🎅 Part 2: {}".format(min(numbers[start:end]) + max(numbers[start:end])))

print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
print("⛄        Day 9         ⛄")
print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()

t0 = time.time()
part1(inputs)
print ("Time: {:.5f}".format(time.time()-t0))

t0 = time.time()
part2(inputs)
print ("Time: {:.5f}".format(time.time()-t0))
